sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -148,6 +148,21 @@ class SimpleTestOptimizer extends Optimizer(
   new SimpleCatalystConf(caseSensitiveAnalysis = true))
 
 /**
+ * Pushes projects down beneath Sample to enable column pruning with sampling.
+ */
+object PushProjectThroughSample extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    // Push down projection into sample
+    case proj @ Project(projectList, Sample(lb, up, replace, seed, child)) =>
+      if (!replace || !projectList.exists(_.find(!_.deterministic).nonEmpty)) {
```

Comment:
The second condition looks complicated. Just `projectList.forall(_.deterministic)`?